try.go

```diff
@@ -72,29 +71,90 @@
 //
 package try
 
+import (
+	"log"
+	"runtime"
+	"testing"
+)
+
 // wrapError wraps an error to ensure that we only recover from errors
 // panicked by this package.
 type wrapError struct{ error }
 
 // Unwrap primarily exists for testing purposes.
 func (e wrapError) Unwrap() error { return e.error }
 
-// Catch catches a previously panicked error and stores it into err.
-// If it successfully catches an error, it calls any provided handlers.
-func Catch(err *error, handlers ...func()) {
-	switch ex := recover().(type) {
+// Recover recovers a previously panicked error and stores it into err.
+// If it successfully recovers an error, and fn is non-nil,
+// it calls fn with the runtime frame in which the error occurred.
+//
+// Recover is a general purpose API.
+// Most use cases will be better served by Handle, HandleF, TB, or Fatal.
+func Recover(errptr *error, fn func(runtime.Frame)) {
+	r(recover(), errptr, fn)
+}
+
+// r implements recover.
+// It is a separate function from Recover to keep stack counts consistent.
+func r(recovered any, err *error, fn func(runtime.Frame)) {
+	switch ex := recovered.(type) {
 	case nil:
 		return
 	case wrapError:
 		*err = ex.error
-		for _, handler := range handlers {
-			handler()
+		if fn != nil {
```

Comment:
Rather than do this dance after recover, I think we should select the appropriate frame at panic time and store it in wrapError. 

I'm not sure if this logic works correctly for multiple rounds of panic-recover-panic-etc.

Reply:
Copy. I was hoping to avoid that cost when not necessary, but it's pretty cheap. Will fix we're ready for another round.